elastic/beats
# [Metricbeat] Update Couchdb to v2

## What does this PR do?

Updates module to use a Strategy patern and decide between v1 and v2 Couchdb parsing. To do so, it has to do a pre HTTP request to root path `/` to know which version is deployed on each fetch (done in fetch to allow hot upgrades in container environment).

I have removed `testdata` tests because one of them was a bit unnecessary with our current `testdata` implementation (a timeout test which was testing `testdata` _framework_ more than the module itself). The other `testdata` doesn't work anymore because more than one request is necessary now to check version.

Because we can only test in CI with a single version, I maintained current Dockerfile

## Why is it important?

ATM, we were only parsing v1 metrics and v3 is going to be released soon. This PR introduces v2 parsing maintaining compatibility with our previous version.

## Checklist

~~- [ ] I have made corresponding change to the default configuration files~~

## How to test this PR locally

* Launch v1 and v2 containers
```
docker run -d --name couchdb16 -p 15984:5984 couchdb:1.6
docker run -d --name couchdb23 -p 25984:5984 -p 25986:5986 couchdb:2.3
```
* Setup Metricbeat Couchdb module to point to both servers `./metricbeat modules enable couchdb` and leave config similar to this:
```
- module: couchdb
  metricsets: ["server"]
  period: 10s
  hosts: ["localhost:15984", "localhost:25984/_node/_local/_stats", "localhost:25984/_node/nonode@nohost/_stats", "localhost:25986"]
```
* Check that events from v1 and v2 are being received in, for example, Kibana. You can see version in field `ecs.service.version` and the `ecs.service.id` must be the same for all paths of v2.
* Extra mile, check dashboards, they must maintain compatibility and show info of both servers.

## Related issues

- Closes elastic/beats#16352
- Relates elastic/beats#10354

## Screenshots
I checked that dashboard still works.

![image](https://user-images.githubusercontent.com/4249331/74947972-edd0c400-53fb-11ea-9326-8f4a1b93e18c.png)
